### PR TITLE
Add tables

### PR DESCRIPTION
**Pull request by schangccg**
for ticket: https://github.com/muccg/bpa-archive-ops/issues/506

### CODE REVIEW

File: bpaotu/bpaotu/query.py [hunk 1]
```diff
@@ -179,6 +179,17 @@ def matching_sample_ids_and_environment(self):
...
+    def matching_sample_headers(self, required_headers=None):
+        query_headers = [SampleContext.id, SampleContext.environment_id]
+
+        if required_headers:
+            for h in required_headers:
+                query_headers.append(getattr(SampleContext, h))
+
+        q = self._session.query(*query_headers)
+
+        return self._q_all_cached('matching_sample_headers', q)
```
Comment (grahame) [inline bpaotu/bpaotu/query.py]: the cache bug is that you're not including required_headers in the cache key; so the key gets set for the first query done, and re-used even if the headers change.

We should sit down together and restructure how the caching code works. A python decorator that marshals function arguments sounds like a nice way. Let's look at this together.

File: bpaotu/bpaotu/templates/bpaotu/tables.html [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+                                    <select class="form-control" id="taxonomy_species"></select>
+                                </td>
+                            </tr>
+                        </tbody>
+                    </table>
+                </div>
+            </div>
+        </div>
+    </div>
+
```
Comment (grahame) [inline bpaotu/bpaotu/templates/bpaotu/tables.html]: is this stuff needed?

File: bpaotu/bpaotu/urls.py [hunk 3]
```diff
@@ -14,4 +14,6 @@
     url(r'^private/api/v1/search$', views.otu_search, name="otu_search"),
     url(r'^private/api/v1/export$', views.otu_export, name="otu_export"),
     url(r'^ingest/$', views.otu_log, name="otu_log"),
+    url(r'^tables/$', views.tables, name="tables"),
+    url(r'^required_table_headers/$', views.required_table_headers, name="required_table_headers"),
```
Comment (grahame) [inline bpaotu/bpaotu/urls.py]: put this under the private API

File: bpaotu/bpaotu/templatetags/html_friendly.py [hunk 4]
```diff
@@ -0,0 +1,9 @@
+from django import template
+
+register = template.Library()
+
+@register.filter
+def html_friendly(str):
+    str = str.replace(" ", "")
+    str = str.lower()
+    return str
```
Comment (grahame) [inline bpaotu/bpaotu/templatetags/html_friendly.py]: use a proper escape method, there's one built into django?

File: bpaotu/bpaotu/views.py [hunk 5]
```diff
@@ -243,6 +247,97 @@ def parse_float(s):
...
+            return datetime.datetime.strptime(s, '%Y-%m-%d').date()
+        except ValueError:
+            return datetime.datetime.strptime(s, '%d/%m/%Y').date()
+
+    def parse_float(s):
+        try:
+            return float(s)
+        except ValueError:
+            return None
+
```
Comment (grahame) [inline bpaotu/bpaotu/views.py]: are these copied out? I think we should move them into a shared file and then import them if so

File: bpaotu/bpaotu/views.py [hunk 6]
```diff
@@ -243,6 +247,97 @@ def parse_float(s):
...
+    results = results[start:start + length]
+
+    def get_environment(environment_id):
+        if environment_id is None:
+            return None
+        return environment_lookup[environment_id]
+
+    data = []
+    for t in results:
+        count = 2
```
Comment (grahame) [inline bpaotu/bpaotu/views.py]: `count = 2` is a bad sign; I think this needs another look :-)

File: bpaotu/bpaotu/templatetags/html_friendly.py (hunk 4 above)
Comment (schangccg) [reply bpaotu/bpaotu/templatetags/html_friendly.py]: Fixed.

File: bpaotu/bpaotu/urls.py (hunk 3 above)
Comment (schangccg) [reply bpaotu/bpaotu/urls.py]: Fixed.

File: bpaotu/bpaotu/views.py (hunk 5 above)
Comment (schangccg) [reply bpaotu/bpaotu/views.py]: Fixed.

File: bpaotu/bpaotu/templates/bpaotu/tables.html (hunk 2 above)
Comment (schangccg) [reply bpaotu/bpaotu/templates/bpaotu/tables.html]: Fixed.

**Comment by grahame** [comment]
Approved pending fix of the cache stuff.

**Pull request merged**